program/rust/src/tests/pyth_simulator.rs

```diff
@@ -65,9 +80,92 @@ impl PythSimulator {
+        }));
+
+
+        let mut program_test = ProgramTest::default();
+        let program_key = Pubkey::new_unique();
+        let programdata_key = Pubkey::new_unique();
+
+        let upgrade_authority_keypair = Keypair::new();
+
+        let program_deserialized = UpgradeableLoaderState::Program {
+            programdata_address: programdata_key,
+        };
+        let programdata_deserialized = UpgradeableLoaderState::ProgramData {
+            slot:                      1,
+            upgrade_authority_address: Some(upgrade_authority_keypair.pubkey()),
+        };
+
+        // Program contains a pointer to progradata
+        let program_vec = bincode::serialize(&program_deserialized).unwrap();
+        // Programdata contains a header and the binary of the program
+        let mut programdata_vec = bincode::serialize(&programdata_deserialized).unwrap();
+        programdata_vec.append(&mut bpf_data);
+
+        let program_account = Account {
+            lamports:   Rent::default().minimum_balance(program_vec.len()),
+            data:       program_vec,
+            owner:      bpf_loader_upgradeable::ID,
+            executable: true,
+            rent_epoch: Epoch::default(),
+        };
+        let programdata_account = Account {
+            lamports:   Rent::default().minimum_balance(programdata_vec.len()),
+            data:       programdata_vec,
+            owner:      bpf_loader_upgradeable::ID,
+            executable: false,
+            rent_epoch: Epoch::default(),
+        };
+
+        // Add to both accounts to program test, now the program is deploy as upgradable
+        program_test.add_account(program_key, program_account);
+        program_test.add_account(programdata_key, programdata_account);
+
+
+        // Start validator
+        let (mut banks_client, payer, mut recent_blockhash) = program_test.start().await;
+
+        // Transfer money to upgrade_authority so it can call the instructions
+        let instruction = system_instruction::transfer(
+            &payer.pubkey(),
+            &upgrade_authority_keypair.pubkey(),
+            1000 * LAMPORTS_PER_SOL,
+        );
+
+        let mut transaction = Transaction::new_with_payer(&[instruction], Some(&payer.pubkey()));
+
+        let blockhash = banks_client
+            .get_new_latest_blockhash(&recent_blockhash)
+            .await
+            .unwrap();
+
+        recent_blockhash = blockhash;
+
+        transaction.partial_sign(&[&payer], recent_blockhash);
+
+        banks_client.process_transaction(transaction).await.unwrap();
+
+        // Same as new() but here we set the payer to be upgrade_authority and program_id to be the
+        // upgradable program
+        return PythSimulator {
+            program_id: program_key,
+            banks_client,
+            payer: upgrade_authority_keypair,
+            last_blockhash: recent_blockhash,
+            programdata_id: Some(programdata_key),
+        };
```

Comment:
i would be in favor of replacing `new` with this method so there's only one simulator setup for tests. I also suggest adding `upgrade_authority` as a separate field to the simulator struct so that users can test with an arbitrary key and the upgrade authority.

Reply:
Good